### PR TITLE
Scriptingv8.1 pull req

### PR DESCRIPTION
**Pull request by mwitchwilliams**
Integrated v8 engine to run JavaScript embedded in X3D Script nodes.

### CODE REVIEW

File: GVRf/Framework/framework/build.gradle [hunk 1]
```diff
@@ -17,7 +17,7 @@ android {
         externalNativeBuild {
```
Comment (liaxim) [inline GVRf/Framework/framework/build.gradle]: There seem to be unnecessary changes to the gradle file.

**Comment by liaxim** [comment]
The v8 .so is missing. It has to be force-added (git add -f) because *.so is in .gitignore.

**Comment by mwitchwilliams** [comment]
Is this possible that these changes were made to make my builds smoother?

I don’t recall making these other than maybe it would help me build.  Really not sure.
Is that something you can over-ride?

Mitch

Mitch Williams
Samsung GearVR X3D development
310-809-4836 (outside line)

**Comment by liaxim** [comment]
Says "This branch cannot be rebased due to conflicts".

**Comment by liaxim** [comment]
Rebased

**Comment by liaxim** [comment]
Note for those who may try this: libjav8.so is under GVRf/framework/framework/src/main/prebuilt/armeabi-v7a/ instead of GVRf/**F**ramework/framework/src/main/prebuilt/armeabi-v7a/. Needs to be copied to the right location.

File: GVRf/Framework/framework/src/main/java/org/gearvrf/script/GVRScriptManager.java [hunk 2]
```diff
@@ -218,7 +219,8 @@ public GVRScriptFile loadScript(GVRAndroidResource resource, String language) th
         if (language.equals(LANG_LUA)) {
             script = new GVRLuaScriptFile(mGvrContext, resource.getStream());
         } else if (language.equals(LANG_JAVASCRIPT)) {
-            script = new GVRJavascriptScriptFile(mGvrContext, resource.getStream());
+            // Commented out for v8ScriptingEngine
+            //script = new GVRJavascriptScriptFile(mGvrContext, resource.getStream());
```
Comment (liaxim) [inline GVRf/Framework/framework/src/main/java/org/gearvrf/script/GVRScriptManager.java]: This breaks the current javascript support in the framework. For example, gvr-javascript crashes since there is no script file set. What is the plan for getting the existing demos working?

Comment (mwitchwilliams) [inline GVRf/Framework/framework/src/main/java/org/gearvrf/script/GVRScriptManager.java]: Initial development, I had to comment out references to the Mozilla Rhino engine to build V8 but still tried to enable Rhino or V8 with a Boolean.  This became impractical and thus created GVRJavascriptV8File.java. to replace GVRJavascriptScriptFile and no longer needed GVRScriptManager.  We could try to run both JavaScript engines  - X3D uses V8, others use Mozilla Rhino.  This might be the fastest way to get this back to working and revisit this issue later.  Otherwise, this seems like a bit of a re-coding effort.

**Comment by thomasflynn** [comment]
when you cherry-pick this, can any sample run?  I seem to get a crash in simplesample about an unsatisfied link error.

**Comment by rahul27** [comment]
Here is @liaxim's comment above. You need to fix the path to the *.so file:

Note for those who may try this: libjav8.so is under GVRf/framework/framework/src/main/prebuilt/armeabi-v7a/ instead of GVRf/Framework/framework/src/main/prebuilt/armeabi-v7a/. Needs to be copied to the right location.

**Comment by thomasflynn** [comment]
sorry, should've mentioned i had already done that.

**Comment by rahul27** [comment]
simplesample works fine for me with this pull request. Could you copy paste the error log?

**Comment by thomasflynn** [comment]
E/AndroidRuntime(20836): FATAL EXCEPTION: main
E/AndroidRuntime(20836): Process: org.gearvrf.simplesample, PID: 20836
E/AndroidRuntime(20836): java.lang.UnsatisfiedLinkError: dlopen failed: unknown reloc type 160 @ 0xe1675164 (33793)
E/AndroidRuntime(20836):        at java.lang.Runtime.loadLibrary(Runtime.java:371)
E/AndroidRuntime(20836):        at java.lang.System.loadLibrary(System.java:988)
E/AndroidRuntime(20836):        at org.gearvrf.GVRActivity.<clinit>(GVRActivity.java:61)
E/AndroidRuntime(20836):        at java.lang.reflect.Constructor.newInstance(Native Method)
E/AndroidRuntime(20836):        at java.lang.Class.newInstance(Class.java:1684)
E/AndroidRuntime(20836):        at android.app.Instrumentation.newActivity(Instrumentation.java:1080)
E/AndroidRuntime(20836):        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2954)
E/AndroidRuntime(20836):        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3218)
E/AndroidRuntime(20836):        at android.app.ActivityThread.access$1000(ActivityThread.java:198)
E/AndroidRuntime(20836):        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1676)
E/AndroidRuntime(20836):        at android.os.Handler.dispatchMessage(Handler.java:102)
E/AndroidRuntime(20836):        at android.os.Looper.loop(Looper.java:145)
E/AndroidRuntime(20836):        at android.app.ActivityThread.main(ActivityThread.java:6837)
E/AndroidRuntime(20836):        at java.lang.reflect.Method.invoke(Native Method)
E/AndroidRuntime(20836):        at java.lang.reflect.Method.invoke(Method.java:372)
E/AndroidRuntime(20836):        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1404)
E/AndroidRuntime(20836):        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1199)

**Comment by liaxim** [comment]
Works for me.

The update of the aars under GearVRf-Demos/gearvrf-libs is not always happening as expected. Couple of things to try.

Run ``./gradlew assembleDebug --rerun-tasks`` in GVRf/Framework. Then build the sample again. Curious to know whether this works out.

Otherwise do a ``git clean -fdx`` in GearVRf and in GearVRf-Demos and then build everything.

**Comment by thomasflynn** [comment]
ok, must be something in my build then.  i'll try again later today.

**Comment by liaxim** [comment]
@mwitchwilliams At a bare minimum this pull request needs corrected location of libjav8.so and uncommented ``script = new GVRJavascriptScriptFile(mGvrContext, resource.getStream());`` in GVRScriptManager.java. Then we can consider merging it after the release.

**Comment by mwitchwilliams** [comment]
SimpleSample worked for me.

Not any JavaScript here so I would not expect the Scripting v8.1 to impact this demo.

Mitch

Mitch Williams
Samsung GearVR X3D development
310-809-4836 (outside line)

**Comment by mwitchwilliams** [comment]
I restored the few lines of code that was left commented out.  However, I still don't get gvr-javascript demo to run.  It crashes in GVRScriptManager.java, bindHelper() function when "_TargetResolver resolver = sBuiltinTargetMap.get(targetName);_" returns a null value for _resolver_.  _targetName_ is set to @GVRScript, but there is no matching value in the TreeMap.  Near the top of GVRScriptManager.java, the values @GVRMain and @GVRActivity are set in the Tree Map, but not @GVRScript. 

Wondering if anyone could test gvr-javascript demo and verify they see the same or validate this demo works, in which case something is wrong with my testing.

Will continue debugging to see if I learn more.

**Comment by mwitchwilliams** [comment]
The files have been reverted (GVRJavascriptScriptFile.java, GVRScriptBehavior.java, GVRScriptFile.java, GVRScriptManager.java and not deleting  jsr223.jar) and the pull request is ready for review.

**Comment by rahul27** [comment]
Since GVRf now supports 64 bit build. This PR needs the 68 bit library for libjav8.so.

**Comment by liaxim** [comment]
Waiting for the 64bit libjav8

**Comment by liaxim** [comment]
Looks like it is ready again for testing and review

**Comment by liaxim** [comment]
Xu has new v8 binaries. @mwitchwilliams will be retesting.

**Comment by mwitchwilliams** [comment]
Retested OK with v8 JavaScript engine update received 4/4/2017

**Comment by liaxim** [comment]
gvr-javascript crashes with:
```
04-19 10:43:44.686 26173 26173 E AndroidRuntime: FATAL EXCEPTION: main
04-19 10:43:44.686 26173 26173 E AndroidRuntime: Process: org.gearvrf.sample.gvrjavascript, PID: 26173
04-19 10:43:44.686 26173 26173 E AndroidRuntime: java.lang.RuntimeException: Unable to start activity ComponentInfo{org.gearvrf.sample.gvrjavascript/org.gearvrf.sample.gvrjavascript.GearVRJavascriptActivity}: java.lang.IllegalArgumentException: Illegal scope value.
04-19 10:43:44.686 26173 26173 E AndroidRuntime: 	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2984)
04-19 10:43:44.686 26173 26173 E AndroidRuntime: 	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3045)
04-19 10:43:44.686 26173 26173 E AndroidRuntime: 	at android.app.ActivityThread.-wrap14(ActivityThread.java)
04-19 10:43:44.686 26173 26173 E AndroidRuntime: 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1642)
04-19 10:43:44.686 26173 26173 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:102)
04-19 10:43:44.686 26173 26173 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:154)
04-19 10:43:44.686 26173 26173 E AndroidRuntime: 	at android.app.ActivityThread.main(ActivityThread.java:6776)
04-19 10:43:44.686 26173 26173 E AndroidRuntime: 	at java.lang.reflect.Method.invoke(Native Method)
04-19 10:43:44.686 26173 26173 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1520)
04-19 10:43:44.686 26173 26173 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1410)
04-19 10:43:44.686 26173 26173 E AndroidRuntime: Caused by: java.lang.IllegalArgumentException: Illegal scope value.
04-19 10:43:44.686 26173 26173 E AndroidRuntime: 	at javax.script.SimpleScriptContext.setAttribute(SimpleScriptContext.java:163)
04-19 10:43:44.686 26173 26173 E AndroidRuntime: 	at org.gearvrf.script.javascript.ExternalScriptable.put(ExternalScriptable.java:184)
04-19 10:43:44.686 26173 26173 E AndroidRuntime: 	at org.gearvrf.script.javascript.RhinoScriptEngine.getRuntimeScope(RhinoScriptEngine.java:309)
04-19 10:43:44.686 26173 26173 E AndroidRuntime: 	at org.gearvrf.script.javascript.RhinoScriptEngine.eval(RhinoScriptEngine.java:157)
04-19 10:43:44.686 26173 26173 E AndroidRuntime: 	at org.gearvrf.script.javascript.RhinoScriptEngine.eval(RhinoScriptEngine.java:199)
04-19 10:43:44.686 26173 26173 E AndroidRuntime: 	at javax.script.AbstractScriptEngine.eval(AbstractScriptEngine.java:224)
04-19 10:43:44.686 26173 26173 E AndroidRuntime: 	at org.gearvrf.script.GVRScriptFile.checkDirty(GVRScriptFile.java:256)
04-19 10:43:44.686 26173 26173 E AndroidRuntime: 	at org.gearvrf.script.GVRScriptFile.invokeFunction(GVRScriptFile.java:175)
04-19 10:43:44.686 26173 26173 E AndroidRuntime: 	at org.gearvrf.script.GVRScriptManager.attachScriptFile(GVRScriptManager.java:187)
04-19 10:43:44.686 26173 26173 E AndroidRuntime: 	at org.gearvrf.script.GVRScriptManager.bindHelper(GVRScriptManager.java:369)
04-19 10:43:44.686 26173 26173 E AndroidRuntime: 	at org.gearvrf.script.GVRScriptManager.bindScriptBundle(GVRScriptManager.java:269)
04-19 10:43:44.686 26173 26173 E AndroidRuntime: 	at org.gearvrf.sample.gvrjavascript.GearVRJavascriptActivity.onCreate(GearVRJavascriptActivity.java:78)
04-19 10:43:44.686 26173 26173 E AndroidRuntime: 	at android.app.Activity.performCreate(Activity.java:6955)
04-19 10:43:44.686 26173 26173 E AndroidRuntime: 	at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1126)
04-19 10:43:44.686 26173 26173 E AndroidRuntime: 	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2927)
04-19 10:43:44.686 26173 26173 E AndroidRuntime: 	... 9 more
```

We can merge this intermediate step but the samples should be still working.

**Comment by liaxim** [comment]
The crash is due to the alternative jsr223 implementation in the v8 jar.

I don't think jsr223 is worth the hassle since it is not official part of the Android libraries. The classes we have for jsr223 have unclear (to me) origin and murky maintenance path. See for example this problem that we have now. We also have to assume we can't redistribute these classes unless we know the license.

We should consider interfacing directly to the Lua and V8 java apis and creating a thin abstraction ourselves, if we must.

**Comment by liaxim** [comment]
Attaching a patch for ExternalScriptable that gets gvr-javascript working with the javax.script classes from the v8 jar.
[diff.txt](https://github.com/Samsung/GearVRf/files/937253/diff.txt)

**Comment by liaxim** [comment]
@mwitchwilliams Could you please talk to Xu and find out whether SimpleScriptContext was something he modified or wrote in the v8 jar. There are few methods (for example getAttributesScope) that return 0 when according to the spec they should be returning -1 in case a variable is not found. Thanks.

**Comment by mwitchwilliams** [comment]
I was just thinking the same thing - figure out what java files changed between jsr223.jar and jav8.jar.
I tested and got the same error.  Due to a build conflict, we removed jsr223.jar since similar files were in jav8.jar. 
OK, let me discuss with Xu further.

**Comment by xcaostagit** [comment]
The change proposed by @liaxim in diff.txt above is probably the safest. Per SimpleScriptContext.java inside jav8.jar, getAttributesScope() may return 3 values: ENGINE_SCOPE(100), GLOBAL_SCOPE(200) or 0 if attribute name is not found. The 0 return is different than what the document says and is probably an error. While we may want to fix the root cause, the change proposed by @liaxim is safer since it doesn't require any more testings.

**Comment by mwitchwilliams** [comment]
Xu, do you want to make the change to jav8.jar and send it to me so I can do some regression tests? And if all is good, I will do the Pull Request with the new jav8.jar.

**Comment by xcaostagit** [comment]
@mwitchwilliams - will do. Hopefully by end of today.

**Comment by mwitchwilliams** [comment]
Retested both gvr-JavaScript and the X3D JavaScript files from GearVRf-Tests/x3d/Scripting_Per_Frame/ repository

**Comment by liaxim** [comment]
Looks good. 64bit builds work too. Planning to merge Thursday morning.

Is there a plan for removing Rhino? Do you think it is something you will get to?

**Comment by mwitchwilliams** [comment]
My understanding is that Rhino will be removed at some point.  V8 had additional demands such as it must be run on the same thread and changed how Bindings (data between Java and JavaScript) were handled.  Thus, it will be some re-coding effort.  Again, my understanding is to come back to this issue, and instead proceed with coding other X3D features.
